account_credit_note_keep_invoiced_quantity/models/account_move.py

```diff
@@ -0,0 +1,26 @@
+# Copyright 2022 Opener B.V. <stefan@opener.amsterdam>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+from odoo import models
+
+
+class AccountMoveLine(models.Model):
+    _inherit = "account.move.line"
+
+    def copy_data(self, default=None):
+        """Circumvent the inclusion of business fields depending on the context.
+
+        This is used to allow the creation of credit notes without resetting the
+        invoiced quantities on sales or purchases.
+        """
+        avoid_include_business_fields = self.env.context.get(
+            "avoid_include_business_fields"
+        ) and self.env.context.get("include_business_fields")
+        if avoid_include_business_fields:
+            self = self.with_context(include_business_fields=False)
+        res = super().copy_data(default=default)
+        if avoid_include_business_fields and "purchase_line_id" in self._fields:
```

Comment:
Hi. maybe not a valid question.
you write "when creating credit notes for sale or purchase invoices" in the doc of the module, but I don't see something for the ``sale_line_id``.
the sale refund is handled by your module ? 
(or maybe I missed something).

Reply:
Good question! We don't have to worry about `sale_line_ids` because it has copy=False: https://github.com/odoo/odoo/blob/14.0/addons/sale/models/account_move.py#L34-L38. That is why they have to be copied in explicitely in the method that my PR avoids: https://github.com/odoo/odoo/blob/14.0/addons/sale/models/account_move.py#L40-L43

But `purchase_line_id` does not have copy=False and is therefore copied by default. There is also an override of the method to copy it explicitely if requested (https://github.com/odoo/odoo/blob/14.0/addons/purchase/models/account_invoice.py#L141-L144), but it is redundant as I found out when testing this.

I consider it a bug in Odoo that `purchase_line_id' is not set to copy=False.